Softeq.XToolkit.Common/Extensions/TaskExtensions.cs

```diff
@@ -11,43 +11,79 @@ namespace Softeq.XToolkit.Common.Extensions
     public static class TaskExtensions
     {
         /// <summary>
-        ///     Task extension to add a timeout.
+        ///     Returns a task that completes as the original task completes or when a timeout expires, whichever happens first.
         /// </summary>
-        /// <returns>The task with timeout.</returns>
-        /// <param name="task">Task.</param>
-        /// <param name="timeout">Timeout.</param>
-        /// <typeparam name="T">Task type.</typeparam>
-        public static async Task<T> WithTimeout<T>(this Task<T> task, TimeSpan timeout)
+        /// <param name="task">The task to wait for.</param>
+        /// <param name="timeout">The maximum time to wait.</param>
+        /// <returns>
+        ///     A task that completes with the result of the specified <paramref name="task"/> or
+        ///     faults with a <see cref="TimeoutException"/> if <paramref name="timeout"/> elapses first.
+        /// </returns>
+        public static async Task WithTimeout(this Task task, TimeSpan timeout)
         {
-            var retTask = await Task.WhenAny(task, Task.Delay((int) timeout.TotalMilliseconds))
-                .ConfigureAwait(false);
+            // Details:
+            //     https://devblogs.microsoft.com/pfxteam/crafting-a-task-timeoutafter-method/
+            // Source:
+            //     https://github.com/microsoft/vs-threading/blob/master/src/Microsoft.VisualStudio.Threading/TplExtensions.cs#L73
 
-            if (retTask is Task<T>)
+            if (task == null)
             {
-                return task.Result;
+                throw new ArgumentNullException(nameof(task));
             }
 
-            return default;
+            using (var timerCancellation = new CancellationTokenSource())
+            {
+                var timeoutTask = Task.Delay(timeout, timerCancellation.Token);
+                var firstCompletedTask = await Task.WhenAny(task, timeoutTask).ConfigureAwait(false);
+
+                if (firstCompletedTask == timeoutTask)
+                {
+                    throw new TimeoutException();
```

Comment:
There's a peculiarity with this case. If the original task will throw an exception - it will never be handled

Reply:
Do we need to handle exceptions if they occur after timeout? Is it dangerous to not handle them?

Reply:
Tasks with unhandled exceptions might cause the following:
"System.AggregateException: A Task's exception(s) were not observed either by Waiting on the Task or accessing its Exception property. As a result, the unobserved exception was rethrown by the finalizer thread."
It's currently more than hundred thousands of such exceptions in HBI's hockeyapp, and the count is rising. Although it's usually not causing the app to actually crash, it floods crash analytics quite a lot.
BTW: the same thing might be caused by exception, thrown in async void method

Reply:
Some words about API backward-compatibility:
- Old impl. of method `WithTimeout` stays and deprecated. (As I can see, don't use anywhere)
- A new method was renamed to `WithTimeoutAsync`.
- Throwing exception it's correct behavior for this method, instead of `null`.

Reply:
I'm not saying throwing exception is not correct. My concern is what gonna happened with original task after exception? It's not completed, and it's abandoned. What if it throw exception?

Reply:
I looked at these cases:
- original Task throws an exception before timeout -> consumer get original exception;
- throws TimeoutException & than original Task throws internally exception -> consumer get TimeoutException, original task will have Exception internally. (How to get task exception, depends on consumer code.)

Reply:
I propose the following solution for the second case:
```cs
var result = task.WithLoggingErrors(logger)
                 .WithTimeoutAsync(timeout);
```